src/testing/cmds/harness/test-collector/main.go

```diff
@@ -0,0 +1,182 @@
+	Output  string
+}
+
+func main() {
+	log.InitPachctlLogger()
+	ctx := pctx.Background("test-collector")
+	tags := flag.String("tags", "", "Tags to run, for example k8s. Tests without this flag will not be selected.")
+	exclusiveTags := flag.Bool("exclusiveTags", true, "If true, ONLY tests with the specified tags will run. If false, "+
+		"the default behavior of 'go test' of including tagged and untagged tests is used.")
+	fileName := flag.String("file", "tests_to_run.csv", "Output file listing the packages and tests to run. Used by the runner script.")
+	pkg := flag.String("pkg", "./...", "Package to run defaults to all packages.")
+	threadPool := flag.Int("threads", 2, "Number of packages to collect tests from concurrently.")
+	flag.Parse()
+	err := run(ctx, *tags, *exclusiveTags, *fileName, *pkg, *threadPool)
+	if err != nil {
+		log.Error(ctx, "Error during tests splitting", zap.Error(err))
+		os.Exit(1)
+	}
+	os.Exit(0)
+}
+
+func run(ctx context.Context, tags string, exclusiveTags bool, fileName string, pkg string, threadPool int) error {
+	tagsArg := ""
+	if tags != "" {
+		tagsArg = fmt.Sprintf("-tags=%s", tags)
+	}
+	testIdsTagged, err := testNames(ctx, pkg, tagsArg)
+	if err != nil {
+		return errors.EnsureStack(err)
+	}
+	var testIds map[string][]string
+	if exclusiveTags && tags != "" {
+		// set difference to get ONLY tagged tests
+		var err error
+		testIdsUntagged, err := testNames(ctx, pkg, "") // collect for set difference
+		if err != nil {
+			return errors.EnsureStack(err)
+		}
+		testIds = subtractTestSet(testIdsTagged, testIdsUntagged)
+	} else {
+		testIds = testIdsTagged
+	}
+	log.Info(ctx, "tests and packages collected", zap.Any("tests", testIds))
+	err = outputToFile(fileName, testIds)
+	return err
+}
+
+// get tests that are tagged, but not in the untagged list since that is inclusive
+func subtractTestSet(testIdsTagged map[string][]string, testIdsUntagged map[string][]string) map[string][]string {
+	resultTests := map[string][]string{}
+	for pkg, testsTagged := range testIdsTagged {
+		testsUntagged, ok := testIdsUntagged[pkg]
+		for _, testNameTagged := range testsTagged {
+			if !ok || !slice.ContainsString(testsUntagged, testNameTagged, func(s string) string { return s }) {
+				if _, ok := resultTests[pkg]; !ok {
+					resultTests[pkg] = []string{testNameTagged}
+				} else {
+					resultTests[pkg] = append(resultTests[pkg], testNameTagged)
+				}
+			}
+		}
+
+	}
+	return resultTests
+}
+
+func testNames(ctx context.Context, pkg string, addtlCmdArgs ...string) (map[string][]string, error) {
+	findTestArgs := append([]string{"test", pkg, "-json", "-list=."}, addtlCmdArgs...)
+	cmd := exec.Command("go", findTestArgs...)
+	cmdReader, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, "GOMAXPROCS=16") // This prevents the command from running wild eating up processes in the pipelines
+	stderr, err := cmd.StderrPipe()
```

Comment:
Does stderr need to be piped anywhere?  I would recommend `cmd.Stderr = log.WriterAt(log.Child(ctx, "stderr"), log.InfoLevel)`